File: cpp/21_search_2d_matrix.cpp
/**
 * LeetCode #74: 搜索二维矩阵
 * 题目: 判断目标值 target 是否存在于按上述规律排序的矩阵中。
 * 解法: 排除法（从右上角出发）- target == matrix[i][j] → 找到；
 *       target < matrix[i][j] → 排除当前列，j--；
 *       target > matrix[i][j] → 排除当前行，i++。
 * 也可二分查找（一维映射）。O(m+n) 或 O(log mn)。
 */

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.empty() || matrix[0].empty()) return false;
        int m = matrix.size(), n = matrix[0].size();
        int i = 0, j = n - 1;
        while (i < m && j >= 0) {
            int cur = matrix[i][j];
            if (cur == target) return true;
            else if (cur < target) ++i;
            else --j;
        }
        return false;
    }
};

int main() {
    Solution sol;
    vector<vector<int>> m = {{1,3,5,7},{10,11,16,20},{23,30,34,60}};
    cout << sol.searchMatrix(m, 3) << endl;  // 1 (true)
    cout << sol.searchMatrix(m, 13) << endl; // 0 (false)
    return 0;
}

File: python/21_search_2d_matrix.py
#!/usr/bin/env python3
"""
LeetCode #74: 搜索二维矩阵
题目: 给定 m×n 矩阵，每行升序，每行第一个元素大于上一行最后一个元素，
      判断目标值 target 是否存在。
解法: 排除法（二分） - 将二维矩阵看成一维有序数组，
      mid = (top + bottom) // 2，比较 matrix[mid//n][mid%n] 与 target。
      时间 O(log(mn))，空间 O(1)。
"""


def searchMatrix(matrix: list[list[int]], target: int) -> bool:
    """
    两次二分或一次二分: 将 2D 当作 1D，用一维的二分下标映射到 2D。
    更直观的"排除法"：从上角（matrix[0][n-1]）开始：
      - target == matrix[i][j] → 找到
      - target < matrix[i][j] → 排除当前列，j--
      - target > matrix[i][j] → 排除当前行，i++
    """
    if not matrix or not matrix[0]:
        return False
    m, n = len(matrix), len(matrix[0])
    i, j = 0, n - 1  # 从右上角出发
    while i < m and j >= 0:
        cur = matrix[i][j]
        if cur == target:
            return True
        elif cur < target:
            i += 1  # 往下走（当前行都太小）
        else:
            j -= 1  # 往左走（当前列都太大）
    return False


def searchMatrix_binary(matrix: list[list[int]], target: int) -> bool:
    """一维二分映射版"""
    if not matrix:
        return False
    m, n = len(matrix), len(matrix[0])
    left, right = 0, m * n - 1
    while left <= right:
        mid = (left + right) // 2
        val = matrix[mid // n][mid % n]
        if val == target:
            return True
        elif val < target:
            left = mid + 1
        else:
            right = mid - 1
    return False


if __name__ == "__main__":
    m = [[1,3,5,7],[10,11,16,20],[23,30,34,60]]
    print(searchMatrix(m, 3))   # True
    print(searchMatrix(m, 13))  # False
